### PR TITLE
fix science transmission for KSP 1.0.5

### LINKED ISSUE #511: Antennas don't transmit science in 1.05 v
Having an issue with the newest version post udate. My space station won't transmit science data. It says transmitting and done, but the science never goes from the station.

### LINKED ISSUE #515: m700 Survey Scanner stuck after 100%
m700 Survey Scanner gets stuck after alleging scan complete, this is tied to incomplete transmission which does not complete scan function. This is needed in order to unlock ore GUI overlay for the celestial body's knowledge base.

### LINKED ISSUE #520: M700 survey dont work
I am trying to survey the mun, but when i click scan, and it counts from 0 to 100%, and i get the msg that its been send, it doesn't show me the overlay map, as it normally would.

It feels like it diden do the survey, or it did not find its way to wherever the data needs to go, for the game to check up on, what i just did.

 if i go to map mode, and focus on the planet, and go to the little menu on the right side, and select the overlay, it tells me "no resource data available"

and ofc its connected to the satellites that is orbiting the kerbin, and as a last try, i also send a maned probe to survey the mun, but with the same result.

I tried reinstall the game, i tried removing all mods (it worked) and i tried removing just the remotetech mod(and it worked) i also made sure i had the latest version of the mod.

### LINKED ISSUE #523: Mobile Processing Lab can't transmit science
- KSP 1.0.5
- RemoteTech 1.6.9
- ScienceAlert *never* installed.

Transmitting science works as expected, except the science stored in an Mobile Processing Lab cannot be transmitted.  When the button is pressed it acts like it is transmitting (counts up to 100%), but the science never appears back at KSC and the science is still present in the MPL.

Cutting this ticket to distinguish between two different bugs being discussed in #511 -- the topic of that one seems to be the ScienceAlert bug where all transmission of science is blocked.  This bug happens without ScienceAlert installed and only affects the MPL.

### PR DESCRIPTION
**Pull request by lamont-granquist**
As reported here:

http://forum.kerbalspaceprogram.com/index.php?/topic/75245-105-remotetech-v169-2015-11-10/&page=150#comment-2285093

Also see, e.g. here:

https://github.com/RemoteTechnologiesGroup/RemoteTech/issues/523#issuecomment-178501552

closes #523
closes #520
closes #515
closes #511

### CODE REVIEW

**Comment by lamont-granquist** [comment]
Note that I'm not setup to build this module right now (I'm on a mac and don't quite grok mono), but this is the patch everyone has been saying they've been applying which fixes these issues.

**Comment by lamont-granquist** [comment]
paging @Peppie23 ?

**Comment by Ghoughpteighbteau** [comment]
+1 for this PR. Seems like a simple fix. 

@lamont-granquist Fancy Mouse said he had to do a silly hack for a NullReferenceException. I don't see that in this pull request, is that relevant?

**Comment by lamont-granquist** [comment]
no don't know anything about any NRE fix...

**Comment by Ghoughpteighbteau** [comment]
haha, wtf.

that's sketchy. Are you just randomly changing things? were you able to identify the NRE?

**Comment by lamont-granquist** [comment]
its got something to do with the last fractional bit of science transmission, so if we just don't do that -- then no NRE -- which is almost certainly wrong, but the bug becomes lolwut rather than game breaking...

File: src/RemoteTech/Modules/ModuleRTDataTransmitter.cs [hunk 1]
```diff
@@ -91,7 +91,7 @@ private IEnumerator Transmit(Callback callback = null)
                 }
```
Comment (langleyVa) [inline src/RemoteTech/Modules/ModuleRTDataTransmitter.cs]: In FancyMouse's fix, line 75-77 changed to:

`               scienceDataQueue.RemoveAt(0);`
`               scienceData.triggered = true;`
`               var subject = ResearchAndDevelopment.GetSubjectByID(scienceData.subjectID);`
`               if (subject == null)`
`                   subject = new ScienceSubject("", "", 1, 0, 0);`
`               int packets = Mathf.CeilToInt(scienceData.dataAmount / PacketSize);`

This seems to be the quirky NRE hack they were talking about.

File: src/RemoteTech/Modules/ModuleRTDataTransmitter.cs [hunk 2]
```diff
@@ -91,7 +91,7 @@ private IEnumerator Transmit(Callback callback = null)
...
                 //StartCoroutine(SetFXModules_Coroutine(modules_progress, 0.0f));
                 float power = 0;
-                while (packets > 0)
```
Comment (langleyVa) [inline src/RemoteTech/Modules/ModuleRTDataTransmitter.cs]: This line is not changed in FancyMouse's fix.

File: src/RemoteTech/Modules/ModuleRTDataTransmitter.cs [hunk 3]
```diff
@@ -122,6 +122,7 @@ private IEnumerator Transmit(Callback callback = null)
                     }
                     yield return new WaitForSeconds(PacketInterval);
                 }
+                GameEvents.OnTriggeredDataTransmission.Fire(scienceData);
```
Comment (langleyVa) [inline src/RemoteTech/Modules/ModuleRTDataTransmitter.cs]: This exact change was done by FancyMouse, too.

**Comment by langleyVa** [comment]
I am currently having the issue with survey data not being transmitted while ScienceAlert is not installed. If this PR fixes the issue, I'd love to see it go through. I diffed FancyMouses version of ModuleRTDataTransmitter.cs to the one in the repo and annotated the PR with what exactly they did.

Sadly, I don't know the code enough to tell whether that is a good fix or not, but hope that collecting that information helps the discussion here.

This is the NRE I found in my KSP.log:

> [LOG 12:51:09.893] Sending data to vessel comms. 1 devices to choose from. Will try to pick the best one
>[ERR 12:51:09.893] [R&D]: No Science Subject found with id survey@Minmus
>[LOG 12:51:11.332] RemoteTech: Changing RnDCommsStream timeout from 0.3 to 0.3200684
>[LOG 12:51:11.332] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 5 - Files to Go: 0
>[LOG 12:51:12.161] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 4 - Files to Go: 0
>[LOG 12:51:12.515] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 3 - Files to Go: 0
>[LOG 12:51:12.862] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 2 - Files to Go: 0
>[LOG 12:51:13.202] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 1 - Files to Go: 0
>[LOG 12:51:13.551] RemoteTech: [Transmitter]: Uploading Data... (Orbital Survey Data) - 6.67 Mits/sec. Packets to go: 0 - Files to Go: 0
>[EXC 12:51:13.551] NullReferenceException: Object reference not set to an instance of an object
	ResearchAndDevelopment.GetScienceValue (Single dataAmount, .ScienceSubject subject, Single xmitScalar)
	ResearchAndDevelopment.SubmitScienceData (Single dataAmount, .ScienceSubject subject, Single xmitScalar, .ProtoVessel source, Boolean reverseEngineered)
	RnDCommsStream.submitStreamData (.ProtoVessel source)
	RnDCommsStream.StreamData (Single dataAmount, .ProtoVessel source)
	RemoteTech.Modules.ModuleRTDataTransmitter+<Transmit>c__Iterator0.MoveNext ()
>[EXC 12:51:13.976] NullReferenceException: Object reference not set to an instance of an object
	ResearchAndDevelopment.GetScienceValue (Single dataAmount, .ScienceSubject subject, Single xmitScalar)
	ResearchAndDevelopment.SubmitScienceData (Single dataAmount, .ScienceSubject subject, Single xmitScalar, .ProtoVessel source, Boolean reverseEngineered)
	RnDCommsStream.submitStreamData (.ProtoVessel source)
	RnDCommsStream+
	.MoveNext ()

**Comment by lamont-granquist** [comment]
certainly looks like a better fix than my hacktastic one.  updated this PR with FancyMouse's version.

**Comment by NathanKell** [comment]
@Peppie23 @erendrake @Starstrider42 anything moving here? :) Or indeed RT generally?

**Comment by lamont-granquist** [comment]
not lately...

**Comment by rsparkyc** [comment]
Please don't tell me RT is dead... Maybe since there was talk about KSP 1.1 adding in these sorts of features (which now has been put off to a later version) this is no longer being maintained?

**Comment by erendrake** [comment]
I hadn't heard that the feature was pushed off 1.1? I should boot RT up and see if we can get some forward motion in here :)

**Comment by lamont-granquist** [comment]
Its also highly likely that the KSP stock version of communication will be a bit dumbed down compared to RT and most likely there will still be RT users that want the old RT experience...

The aero model in stock didn't make FAR go away, and the resource model in stock didn't make Karbonite/Regolith go away.

Be super nice to get this bug fixed one way or another and make a patch release.

**Comment by rsparkyc** [comment]
I agree, I don't recall if the stock version will have signal delay.  [This](http://wiki.kerbalspaceprogram.com/wiki/Planned_features) shows relay networks being delayed till 1.2, which points to [this](http://forum.kerbalspaceprogram.com/index.php?/topic/131794-devnote-tuesday-wednesday-edition-ii/) blog post, but now that i'm re-reading the post I can't see where that's being delayed.

**Comment by lamont-granquist** [comment]
I know for certain I read it in one of the devnotes tuesday-wednesdays that it was cut out of 1.1

**Comment by lamont-granquist** [comment]
Here found the right thread with the announcement that new comms had been cut from 1.1:

http://forum.kerbalspaceprogram.com/index.php?/topic/131399-no-new-comms-system-in-11/

**Comment by ArSn** [comment]
@NathanKell There is not much going on here atm indeed.

Granted, I did not contribute a lot to this project (yet), but I feel like I should give my 2 cents in the discussion here of why that is.

I had a lot of fun working on the issues that were around when there was no word at all of the relay system being integrated in stock KSP. However, as some people here already suggested, the notice of this system being added in 1.1 really took the wind out of our sails (or at least mine, that is).
It is unfortunate that it has been delayed to after 1.1, but that decision is not too old now.

Still, the problem at hand is, that one just does not know how much of it will be integrated in stock, and therefore, one does also not know if RT will be dead after the update which includes the relay system in stock (whether that is 1.2, 1.3 or anything else).

I would be happy to spin up my efforts in this project again, if there is actually an audience that likes to keep it, and also, if it will actually still make a difference to the stock game. I know you are not a psychic either, and thus can not promise what the future will bring, however, for the people contributing to RT, I imagine it to be crucial information, whether, when, and how much of RT is going to be integrated in stock, and thus, I would highly appreciate it if there would be a more in-detail explanation of those plans.

**Comment by NathanKell** [comment]
@ArSn the stock feature is not going to be RT-like, AFAIK, except to the extent that relay networks are involved. There certainly won't be a flight computer, and probably not signal delay either; my guess is just that certain functions (but not all functions) will be locked out when off the grid. (I've been busy in my own dev work, haven't had a chance to play at all with that branch; though NDA would not let me say much if I had, either, sadly).

I can state for a fact that Realism Overhaul (and its entire user base) will still be wanting to use RemoteTech going forward, and my guess is many others will as well ( @lamont-granquist 's comparison to stock aero vs FAR will be very apt, I think).

**Comment by Ghoughpteighbteau** [comment]
I think if you're going to be putting your money on what the future will look like, then put your money on stock communications being as friendly as possible. That has been the Squad modus operandi for everything. Planes that worked in the updated aerodynamics still flew into the ground in FAR. Remote tech is a cruel cruel mod, an aspect Squad will never copy. There's always going to be a market for masochism in KSP. So... I wouldn't worry about it.

Do you need some third parties to test this PR? I can get a dev environment set up if need be.

**Comment by langleyVa** [comment]
@BobPalmer, you implement the stock relay network features, right? Can you drop a few words on whether RemoteTech will still have it's place or nah?

**Comment by Starstrider42** [comment]
Maybe I should say something since I got @mentioned. :wink: 

I think the main problem has been that none of the developers have been available for the last few months. I'm finding it hard just to keep my own mod going (this week marked the first time in *two months* that I did any KSP modding), let alone a project as demanding as RemoteTech. I'm pretty sure that Peppie23 is in the same ignore-everything-KSP limbo that I spent a year in, as I tried contacting him when I got back but never got a reply.

That said, I was pleasantly surprised to see Erendrake's response above (and I have no idea who ArSn is, but belated welcome to the RTG). Hopefully they'll be able to fix at least the most frustrating bugs. I'm afraid I can't commit even to taking up my old job with the RT manual, let alone actual development. I'll let you guys know if things change.

**Comment by Ghoughpteighbteau** [comment]
If the current remote tech devs are not feeling up to maintaining, that's *fine*! You're not slaves, but you should consider promoting some enthusiastic fans. Take advantage of the good old open source model.

**Comment by tomekpiotrowski** [comment]
@d4rksh4de Is this whole science transmission issue (reported here #511) still a problem in KSP 1.1?

**Comment by d4rksh4de** [comment]
@tomekpiotrowski Haven't checked that.

**Comment by tomekpiotrowski** [comment]
I will add this to the milestone for now. If it was an issue in 1.0.5 there's a good chance it's still an issue now.

**Pull request merged**